src/main/java/games/strategy/engine/framework/headlessGameServer/AvailableGames.java

```diff
@@ -75,11 +78,28 @@ String getGameFilePath(final String gameName) {
   private static void populateAvailableGames(final Map<String, URI> availableGames,
       final Set<String> availableMapFolderOrZipNames, final Set<String> mapNamePropertyList) {
     System.out.println("Parsing all available games (this could take a while). ");
-    for (final File map : allMapFiles()) {
-      if (map.isDirectory()) {
-        populateFromDirectory(map, availableGames, availableMapFolderOrZipNames, mapNamePropertyList);
-      } else if (map.isFile() && map.getName().toLowerCase().endsWith(ZIP_EXTENSION)) {
-        populateFromZip(map, availableGames, availableMapFolderOrZipNames, mapNamePropertyList);
+    final List<File> files = allMapFiles();
+    if (files.size() > 0) {
+      final ExecutorService service = Executors.newWorkStealingPool(files.size());
+      final List<Future<?>> tasks = new ArrayList<>(files.size());
+      for (final File map : files) {
+        if (map.isDirectory()) {
+          tasks.add(service.submit(
+              () -> populateFromDirectory(map, availableGames, availableMapFolderOrZipNames, mapNamePropertyList)));
+        } else if (map.isFile() && map.getName().toLowerCase().endsWith(ZIP_EXTENSION)) {
+          tasks.add(service.submit(
+              () -> populateFromZip(map, availableGames, availableMapFolderOrZipNames, mapNamePropertyList)));
+        }
+      }
+      service.shutdown();
+      for (Future<?> future : tasks) {
```

Comment:
Any reason why you can't just call `service.awaitTermination()` here?  It doesn't appear that you are using any of the task results and are simply using them to join the current thread on the worker threads, which `awaitTermination()` should do, as well.

Alternatively, since this is a one-off executor service anyway, you could just submit all the tasks at once using `invokeAll()`, which doesn't return until all tasks are complete, and avoid `shutdown()`, `awaitTermination()`, `get()`, etc.

Reply:
awaitTermination requires a timeout parameter, which would then need to be set to an arbitrary high value with no real reasoning.
Using invokeAll is probably the best idea.

Reply:
Agreed that `invokeAll()` is probably the better choice here.

FWIW, the Javadocs for the [`java.util.concurrent`](https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/package-summary.html) package states:

>  All methods that accept timeout parameters treat values less than or equal to zero to mean not to wait at all. **To wait "forever", you can use a value of `Long.MAX_VALUE`.**

It's always nice to get an official blessing from Oracle on what would normally seem like a hack. :smile:

Reply:
Agreed 👍